clang/lib/CIR/Dialect/Transforms/TargetLowering/LowerModule.h

```diff
@@ -0,0 +1,82 @@
+//===--- LowerModule.h - Abstracts CIR's module lowering --------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file partially mimics clang/lib/CodeGen/CodeGenModule.h. The queries are
+// adapted to operate on the CIR dialect, however.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_CIR_DIALECT_TRANSFORMS_TARGETLOWERING_LOWERMODULE_H
+#define LLVM_CLANG_LIB_CIR_DIALECT_TRANSFORMS_TARGETLOWERING_LOWERMODULE_H
+
+#include "CIRContext.h"
+#include "LowerTypes.h"
+#include "MissingFeature.h"
+#include "TargetLoweringInfo.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/DataLayoutInterfaces.h"
+#include "clang/Basic/TargetInfo.h"
+#include "clang/CIR/Dialect/IR/CIRDialect.h"
+
+namespace mlir {
+namespace cir {
+
+/// Replaces CodeGenModule from Clang in ABI lowering.
+class LowerModule {
```

Comment:
Instead of replacing, I'd prefer if `LowerModule` would become a `CIRModule` thing that is instantiated as part of `CodeGenModule`. The ownership of `CIRModule` can just be transferred down to target lowering.

Reply:
I would rather remove this class altogehter instead of adding this `CIRModule`. This class doesn't provide much for the lowering, it was only added to make the codegen parity more obvious (see comment below).

Reply:
I'll remove it :laughing:

If we see a more compelling use case for this abstraction we can add it later.

Reply:
@bcardosolopes I'm going back on my decision to remove `LowerModule`.

I found it more useful than I initially thought. It aggregates a bunch of data required to initialize ABI stuff, making it easy to pass this info to functions.

But I'm against raising it or creating a `CIRModule`:

 - It does not intersect much with `CIRGenModule` despite being inspired by it.
 - Lowering prepare already has access to this class.
 
Given the above, can I keep it as is?

Reply:
> I'm going back on my decision to remove LowerModule.

Ok.